### PR TITLE
[Cache] Unconditionally use PhpFilesAdapter for system pools

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Now that we're about to leverage OPCache shared memory even for objects (see #27543), there's no reason anymore to use APCu for system caches. Let's remove some complexity here.

As a bonus, this makes system caches pruneable using the `cache:pool:prune` command.

### CODE REVIEW

**Comment by palex-fpt** [comment]
There is one problem with PhpFilesAdapter - it does not use opcache till next request. 
```
 $cache->set('key', 'value');
 $cache->get('key'); // load directly from file, as file with timestamp >= request start time is not opcached.

**Comment by nicolas-grekas** [comment]
> There is one problem with PhpFilesAdapter - it does not use opcache till next request.

After having a deeper look, it does, but only if the file's mtime is in the past. This is now fixed. Thanks for the hint.

**Comment by palex-fpt** [comment]
to pass tests it should invalidate cache of file deletion (opcache_invalidate should be called before file is unlinked)

**Comment by nicolas-grekas** [comment]
> to pass tests it should invalidate cache on file deletion (opcache_invalidate should be called before file is unlinked)

That would open race conditions (invalidating, then a concurrent reloads the legacy file, the we write but this is ignored). Tests do pass, so not sure what you meant :)

**Comment by palex-fpt** [comment]
calls to $cache->delete() or $cache->clear() unlinks files. but opcache is still has file cached. And `list($expiresAt, $values[$id]) = include $file;` returns data. Tests pass when opcache is not working (opcache_cli = false)

**Comment by palex-fpt** [comment]
```
Index: src/Symfony/Component/Cache/Traits/PhpFilesTrait.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/Symfony/Component/Cache/Traits/PhpFilesTrait.php	(revision 3a2eb0d9514f0b322ab31a128aa48b9f89e28d50)
+++ src/Symfony/Component/Cache/Traits/PhpFilesTrait.php	(date 1528439846061)
@@ -23,6 +23,9 @@
  */
 trait PhpFilesTrait
 {
+    use FilesystemCommonTrait {
+        doUnlink as doFsUnlink;
+    }
     use FilesystemCommonTrait;
 
     private $includeHandler;
@@ -155,4 +158,13 @@
 
         return $ok;
     }
+
+    protected function doUnlink($file)
+    {
+        $allowCompile = 'cli' !== PHP_SAPI || ini_get('opcache.enable_cli');
+        if ($allowCompile) {
+            @opcache_invalidate($file, true);
+        }
+        return $this->doFsUnlink($file);
+    }
 }
Index: src/Symfony/Component/Cache/Traits/FilesystemCommonTrait.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/Symfony/Component/Cache/Traits/FilesystemCommonTrait.php	(revision 3a2eb0d9514f0b322ab31a128aa48b9f89e28d50)
+++ src/Symfony/Component/Cache/Traits/FilesystemCommonTrait.php	(date 1528439709379)
@@ -56,7 +56,7 @@
         $ok = true;
 
         foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS)) as $file) {
-            $ok = ($file->isDir() || @unlink($file) || !file_exists($file)) && $ok;
+            $ok = ($file->isDir() || @$this->doUnlink($file) || !file_exists($file)) && $ok;
         }
 
         return $ok;
@@ -71,7 +71,7 @@
 
         foreach ($ids as $id) {
             $file = $this->getFile($id);
-            $ok = (!file_exists($file) || @unlink($file) || !file_exists($file)) && $ok;
+            $ok = (!file_exists($file) || @$this->doUnlink($file) || !file_exists($file)) && $ok;
         }
 
         return $ok;
@@ -122,7 +122,12 @@
             parent::__destruct();
         }
         if (null !== $this->tmp && file_exists($this->tmp)) {
-            unlink($this->tmp);
+            $this->doUnlink($this->tmp);
         }
     }
+
+    protected function doUnlink($file)
+    {
+        return @unlink($file);
+    }
 }
```

**Comment by nicolas-grekas** [comment]
Thanks, patch applied.

**Comment by palex-fpt** [comment]
opcache_invalidate should be called before unlink

**Comment by palex-fpt** [comment]
invalidate fails when file does not exits and opcache still contains file entry.

**Comment by nicolas-grekas** [comment]
Thanks, now green!

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 1]
```diff
@@ -43,7 +43,6 @@ public function __construct(string $file, AdapterInterface $fallbackPool)
     {
         $this->file = $file;
         $this->pool = $fallbackPool;
-        $this->zendDetectUnicode = ini_get('zend.detect_unicode');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: Dealing with zend unicode anymore when [this line](https://github.com/symfony/symfony/pull/27549/files#diff-8e644226ec6cb06d1d233ff5a4d1e4e3L101) is removed. Let's remove the related complexity and overhead.

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 2]
```diff
@@ -87,9 +87,13 @@ function ($deferred, $namespace, &$expiredIds) use ($getId) {
      * @param LoggerInterface|null $logger
      *
      * @return AdapterInterface
+     *
+     * @deprecated since Symfony 4.2.
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: no dot at the end (I know, our conventions are hard to follow :)).

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: updated

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 3]
```diff
@@ -102,9 +102,13 @@ function ($deferred, $namespace, &$expiredIds) use ($getId) {
      * @param LoggerInterface|null $logger
      *
      * @return AdapterInterface
+     *
+     * @deprecated since Symfony 4.2
      */
     public static function createSystemCache($namespace, $defaultLifetime, $version, $directory, LoggerInterface $logger = null)
     {
+        @trigger_error(sprintf('The "%s()" method is deprecated since Symfony 4.2.', __CLASS__), E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: shouldn't it be `__METHOD__` ?
